### ISSUE TITLE
checkURL: detect IPv4-non-decimal-part validation error (hex/octal host notation)

### ISSUE THREAD

**Issue by YusukeHirao** [markuplint/markuplint#3966]
## Summary

`checkURL` (`packages/@markuplint/types/src/whatwg/check-url.ts`) delegates structural URL parsing to `URL.canParse()` and layers regex-based detection on top for URL LS validation errors that `new URL()` silently auto-corrects (invalid-credentials, special-scheme-missing-following-solidus, file-invalid-Windows-drive-letter, etc. — see the file's own doc comment for the full list). It has no detection for `IPv4-non-decimal-part`: a host label written in hex (`0x...`) or octal (leading-zero decimal) notation. `URL.canParse()` / `new URL()` accept and silently normalize these (Node's URL implementation follows the URL LS "IPv4 number parser" algorithm, which treats hex/octal as a *validation error*, not a parse failure), so markuplint currently reports 0 violations for such hosts.

This surfaced as a **new** `nu-only` regression after a routine `yarn bench:update` picked up a newer `ghcr.io/validator/validator` image digest — the nu-validator image apparently gained (or fixed) this check recently. It is unrelated to the `nu-only` backlog this Issue's sibling Issues (#3963, #3964, #3965) were filed against.

## Repro

```html
<base href="http://192.0x00A80001">
```
```html
<base href="http://０Ｘｃ０．０２５０．０１">
```

```html
<base href="http://%30%78%63%30%2e%30%32%35%30.01">
```

nu-validator (current image): error — `Invalid host: IPv4 address contains a non-decimal or leading-zero part.` for all three.

markuplint (all bench rules enabled): 0 violations for all three. Confirmed empirically: `URL.canParse('http://192.0x00A80001')` → `true`, `new URL('http://192.0x00A80001').host` → `'192.168.0.1'` (Node silently accepts and normalizes the hex notation).

Fixtures (newly `nu-only` in `tests/external/snapshots/diff/nu-only.json`):
- `tests/external/validator/tests/html/elements/base/href/host-192.0x00A80001-isvalid.html`
- `tests/external/validator/tests/html/elements/base/href/host-IP-address-fullwidth-isvalid.html`
- `tests/external/validator/tests/html/elements/base/href/host-IP-address-percent-encoded-isvalid.html`

Note: despite the `-isvalid` filename suffix (the validator/validator test corpus's own convention marking a fixture as testing spec-valid markup), the *current* URL Living Standard treats this host notation as carrying a validation error — see below. The corpus fixture name likely predates the "IPv4-non-decimal-part" wording landing in the spec, or predates nu-validator implementing it.

## Spec basis

[URL Standard §3.5 "Host parsing", the IPv4 number parser](https://url.spec.whatwg.org/#ipv4-number-parser):

> If input contains at least two code points and the first two code points are either "0X" or "0x", then: Set validationError to true. Remove the first two code points from input. Set R to 16.
>
> Otherwise, if input contains at least two code points and the first code point is U+0030 (0), then: Set validationError to true. Remove the first code point from input. Set R to 8.

And the [IPv4 parser](https://url.spec.whatwg.org/#concept-ipv4-parser) that calls it, per dot-separated part:

> Let result be the result of parsing part. If result is failure, IPv4-non-numeric-part validation error, return failure. If result[1] is true, IPv4-non-decimal-part validation error.

Crucially, `IPv4-non-decimal-part` is a **validation error**, not a parse failure — the URL still parses successfully (which is why `URL.canParse()` correctly returns `true` and Node's behavior is not a bug). It is exactly the same class of "parses fine, but is a URL LS conformance violation" issue that `checkURL` already handles for `invalid-credentials`, `special-scheme-missing-following-solidus`, etc.

## Proposed direction

Add a check to `checkURL` (`packages/@markuplint/types/src/whatwg/check-url.ts`) for the host component:

1. Extract the host from the (already-parsed, via `new URL(trimmed, ...)`) authority — only meaningful for special-scheme URLs where IPv4 parsing applies.
2. Split the host on `.`, and for each dot-separated label, check whether it matches `^0[xX]` (hex) or `^0\d` (leading-zero decimal) *before* percent-decoding/IDNA processing is done by `new URL()` — since `new URL()` already normalizes the host (e.g., to `192.168.0.1`), checking the **post-parse** host will miss the hex/octal notation entirely. The check likely needs to run against a pre-normalized view of the host label, or against `url.hostname` compared to see if any numeric-looking segment was reinterpreted.
3. Handle the fullwidth-Unicode and percent-encoded obfuscation variants — both need their own decode/normalize step (NFKC + percent-decode) before the hex/octal check applies, mirroring what the URL parser's own IDNA/percent-decode preprocessing does internally. This is the trickiest part of the implementation: naively checking `trimmed` for `0x`/leading-zero substrings would produce false positives on paths/queries containing coincidental digit sequences, so the check must be scoped precisely to the host component after the same preprocessing the spec's host parser performs.

This is more involved than the other regex-based checks already in the file (most of which operate on the raw pre-parse string) because it requires partially replicating host-parsing preprocessing to reliably isolate hex/octal notation before `new URL()` normalizes it away. Recommend a proof-of-concept against the three repro cases above before committing to a general approach.

## Files

- `packages/@markuplint/types/src/whatwg/check-url.ts`
- `packages/@markuplint/types/src/whatwg/check-url.spec.ts`

## Bench impact

3 fixtures currently `nu-only` (surfaced by a nu-validator image update, not part of the original triage backlog). Resolving this Issue is expected to flip them to `match-error`.

## Benchmark cross-reference

_Sync'd from `tests/external/snapshots/` (submodule=1429313, nu-validator@sha256:210bbc353b23, markuplint=5.0.0-rc.4, generated=2026-08-12 06:31:53 UTC)._

Filter: `^html\/elements\/base\/href\/host-(192\.0x00A80001|IP-address-fullwidth|IP-address-percent-encoded)-isvalid` → **3** fixtures. Verdict tally: match-error=0, match-clean=0, ml-only=0, nu-only=3, nu-over=0.

[URL Standard §3.5 "Host parsing", IPv4 number parser](https://url.spec.whatwg.org/#ipv4-number-parser): a dot-separated host label starting with "0x"/"0X" (hex) or a leading "0" (octal) sets `validationError` to true — `IPv4-non-decimal-part` — even though the number still parses successfully. `checkURL` (`packages/@markuplint/types/src/whatwg/check-url.ts`) delegates structural parsing to `URL.canParse()`/`new URL()`, which silently normalizes hex/octal notation (verified: `new URL('http://192.0x00A80001').host` → `'192.168.0.1'`) without surfacing the validation error, matching the pattern already handled for `invalid-credentials`/`special-scheme-missing-following-solidus`/etc. in the same file. Surfaced by a nu-validator Docker image update (unrelated to the original nu-only backlog these fixtures' `-isvalid` naming predates the URL LS wording landing/nu implementing it). Needs host-component isolation plus percent-decode/NFKC-normalize preprocessing before the hex/octal check, to handle the fullwidth-Unicode and percent-encoded obfuscation variants.

| fixture | verdict | nu id / ml rule |
| --- | --- | --- |
| `html/elements/base/href/host-192.0x00A80001-isvalid.html` | nu-only | `nv-408800eea8f3` |
| `html/elements/base/href/host-IP-address-fullwidth-isvalid.html` | nu-only | `nv-6f8e202afcff` |
| `html/elements/base/href/host-IP-address-percent-encoded-isvalid.html` | nu-only | `nv-51ed8390068d` |

Reproduce: see [`bench-triage` skill](https://github.com/markuplint/markuplint/blob/dev/.claude/skills/bench-triage/SKILL.md) for the audit workflow.
